HTTPS & TLS docs (#1341)
microsoft/reverse-proxy: docs/docfx/articles/https-tls.md

---
uid: https-tls
title: HTTPS & TLS

> ---

Reviewer: Nit: We seem to be inconsistent with regards to including this header on articles. Does it improve something with docfx?

Author: No idea 😁, I copied it from another doc. I'll drop it until we find a use for them.

# HTTPS & TLS

> HTTPS (HTTP over TLS encrypted connections) is the standard way to make HTTP request on the Internet for security, integrity, and privacy reasons. There are several HTTPS/TLS considerations to account for when using a reverse proxy like YARP.

Reviewer: ```suggestion
HTTPS (HTTP over TLS encrypted connections) is the standard way to make HTTP requests on the Internet for security, integrity, and privacy reasons. There are several HTTPS/TLS considerations to account for when using a reverse proxy like YARP.
```

## TLS Termination

YARP is a level 7 HTTP proxy which means that incoming HTTPS/TLS connections are fully decrypted by the proxy so it can process and forward the HTTP requests. This is commonly known as TLS Termination. The outgoing connections to the destination(s) may or may not be encrypted, depending on the configuration provided.

> ## TLS tunnelling (CONNECT)

Reviewer 2: ```suggestion
## TLS tunneling (CONNECT)
```

TLS tunneling using the CONNECT method is a feature used to proxy requests without decrypting them. This is _not_ supported by YARP and there are no plans to add it.

## Configuring incomming connections

YARP can run on top of all ASP.NET Core servers and configuring HTTPS/TLS for incoming connections is server specific. Check the docs for [Kestrel](https://docs.microsoft.com/aspnet/core/fundamentals/servers/kestrel/endpoints#listenoptionsusehttps), [IIS](https://docs.microsoft.com/iis/manage/configuring-security/how-to-set-up-ssl-on-iis), and [Http.Sys](https://docs.microsoft.com/aspnet/core/fundamentals/servers/httpsys#configure-windows-server-1) for configuration details.

### Advanced TLS filters with Kestrel

> Kestrel supports intercepting incoming connections before the TLS handshake. YARP includes a [TlsFrameHelper](xref:Yarp.ReverseProxy.Utilities.Tls.TlsFrameHelper) API that can parse the raw TLS handshake and enable you to gather custom telemetry or eagerly reject connections. These APIs cannot modify the TLS handshake or decrypt the data stream. See this [example](https://github.com/microsoft/reverse-proxy/blob/v1.0.0-rc.1/testassets/ReverseProxy.Direct/TlsFilter.cs).

Reviewer: ```suggestion
Kestrel supports intercepting incoming connections before the TLS handshake. YARP includes a [TlsFrameHelper](xref:Yarp.ReverseProxy.Utilities.Tls.TlsFrameHelper) API that can parse the raw TLS handshake and enable you to gather custom telemetry or eagerly reject connections. These APIs cannot modify the TLS handshake or decrypt the data stream. See this [example](https://github.com/microsoft/reverse-proxy/blob/main/testassets/ReverseProxy.Direct/TlsFilter.cs).
```

Nit: link to main branch should be less likely to die since we remove old release branches.

Author: I didn't want the link to break if we moved the code. Note I've linked to the tag rather than the branch so it shouldn't get deleted.

Reviewer: Oh I missed that it's the tag 👍
